### PR TITLE
Improve code coverage numbers by excluding some options

### PR DESCRIPTION
**Pull request by LarryOsterman**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: cmake-modules/AzureCodeCoverage.cmake [hunk 1]
```diff
@@ -55,7 +55,8 @@ function(add_gcovr_html)
             -r ${BASEDIR} ${EXCLUDE_ARGS}
             --object-directory=${PROJECT_BINARY_DIR}
             -o ${args_TARGET_NAME}/index.html
-
+            --exclude-unreachable-branches
```
Comment (RickWinter) [inline cmake-modules/AzureCodeCoverage.cmake:58]: These would be dead code? Would be great if we could remove em

File: cmake-modules/AzureCodeCoverage.cmake [hunk 2]
```diff
@@ -55,7 +55,8 @@ function(add_gcovr_html)
...
+            --exclude-throw-branches
```
Comment (RickWinter) [inline cmake-modules/AzureCodeCoverage.cmake:59]: These we should leave and add tests that will throw from those paths.

**Review by RickWinter — COMMENTED** [review]
Removing the unreachable is reasonable. (Could we remove the dead code??)

Lets keep the throw paths as those should be coverable by adding tests that throw from the paths. Granted it might be hard to add those tests, but the branches are valid.

File: cmake-modules/AzureCodeCoverage.cmake (hunk 2 above)
Comment (LarryOsterman) [reply cmake-modules/AzureCodeCoverage.cmake:59]: "throw branches" are not "branches that contain throw directives". "throw branches" are branches which the compiler generates whenever you call a method which may throw.

So it's not feasible to write code to cover those branches, unless we figure out a way to convince the C++ standard libraries to throw.

There's more information in the [GCovr FAQ](https://gcovr.com/en/stable/faq.html#why-does-c-code-have-so-many-uncovered-branches)

File: cmake-modules/AzureCodeCoverage.cmake (hunk 1 above)
Comment (LarryOsterman) [reply cmake-modules/AzureCodeCoverage.cmake:58]: These are *not* dead code. These are code branches generated by the compiler which can never be executed.

There's more information in the FAQ about them (https://gcovr.com/en/stable/faq.html#why-does-c-code-have-so-many-uncovered-branches).

Comment (ahsonkhan) [reply cmake-modules/AzureCodeCoverage.cmake:58]: Maybe it's worth adding the link to this doc as a comment in this source file, for later reference.

**Comment by ahsonkhan** [comment]
Overall branch coverage went up from 53% to 69% for core.

Can we bump this up to 65%?
https://github.com/Azure/azure-sdk-for-cpp/blob/b9c776c9ff2e376a71c42d6db34cbfa0184240bb/sdk/core/ci.yml#L55

As an example: This change just improved the branch coverage numbers for `url_cpp` from 70% to 95%, and more accurately reflects the coverage of "our code" with the tests we have. The great thing about that is, the remaining 5% branch that is not covered, is actually actionable and we can improve. And that's now highlighted where it was previously lost in the noise! Thanks!

A lot of the partial branch coverage previously showing for using string literals, ctor, and concat that are unactionable for us to write tests against, are gone.

[Before](https://dev.azure.com/azure-sdk/public/_build/results?buildId=3968241&view=codecoverage-tab)
image

[After](https://dev.azure.com/azure-sdk/public/_build/results?buildId=3967618&view=codecoverage-tab)
image

File: cmake-modules/AzureCodeCoverage.cmake (hunk 2 above)
Comment (ahsonkhan) [reply cmake-modules/AzureCodeCoverage.cmake:59]: > With the [gcovr --exclude-throw-branches](https://gcovr.com/en/stable/manpage.html#cmdoption-gcovr-exclude-throw-branches) option, exception-only branches will be ignored. These are typically arcs from a function call into an exception handler.

> For branch coverage, exclude branches that the compiler generates for exception handling. This often leads to more ‘sensible’ coverage reports. Config key(s): exclude-throw-branches.



Interesting!

**Review by ahsonkhan — APPROVED** [review]

**Comment by ahsonkhan** [comment]
From a quick scan of some of the files, I don't see any issues with false reporting of "coverage" which aren't. This change is great overall, are we reasonably confident we don't have false coverage for some cases? Given the documentation, and what I saw, it seems reasonable. We could keep an eye out on new code changes, if we detect higher coverage numbers than we'd expect.

Should we also add `--exclude-noncode-lines`? Maybe it isn't relevant in our code base to make a difference, unlike the other two options.

File: cmake-modules/AzureCodeCoverage.cmake (hunk 1 above)
Comment (LarryOsterman) [reply cmake-modules/AzureCodeCoverage.cmake:62]: A good idea, done.

**Pull request merged**